_posts/2021-04-05-reactive-mode.md

```diff
@@ -0,0 +1,150 @@
+- The **TaskManager(s)** are deployed as a Kubernetes deployment, which is scaled through a [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). In this experiment, the autoscaler is monitoring the CPU load of the pods in the deployment. The number of pods is adjusted between 1 and 15 pods by the autoscaler.
+
+**Additional Components**:
+
+- We have a **Zookeeper** and **Kafka** deployment (each with one pod) to provide a Kafka topic that serves as the input for the Flink job.
+- **Data Generator** pod which produces messages to the Kafka topic. The data generator is generating simple string messages at a adjustable rate. In this experiment, the rate is following a sine wave.
+- For monitoring, we are deploying **Prometheus** and **Grafana**.
+
+The entire setup is [available on GitHub](https://github.com/rmetzger/flink-reactive-mode-k8s-demo) if you want to try this out yourself.
+
+### Results
+
+We've deployed above components on a hosted Kubernetes cluster, running it for several days. The results are best examined based on a simple Grafana dashboard:
+
+<div class="row front-graphic">
+  <img src="{{ site.baseurl }}/img/blog/2021-04-reactive-mode/result.png" alt="Reactive Mode Demo Result"/>
+  <p class="align-center">Reactive Mode Experiment Results</p>
+</div>
+
+Let's take a look at the dashboard:
+
+- On the top left, we see the **Kafka consumer lag**, reported by Flink's Kafka consumer (source). This metric reports basically the queue size of unprocessed messages. A high lag means that Flink is not processing messages as fast as they are produced: we need to scale up. 
+
+  In the chart you see that the lag is usually following the throughput of data coming from Kafka. When the throughput is the highest, the reported lag is at \~75k messages. In low throughput times, it is basically at zero.
+
+- On the top right, you see the **throughput**, measured in Records per second, as reported by Flink. You see that the throughput is roughly following a sine wave, peaking at 6k messages per second, and going down to almost zero.
+
+- The bottom left chart is showing the **CPU load** per TaskManager. We've added this metric to the dashboard, since this is the metric used by Kubernetes to decide on the replica count of the TaskManager deployment. We can see that as soon as a certain CPU load is reached, additional TaskManagers are started.
+
+- The bottom right chart is showing the **TaskManager count** over time. We see when throughput (and CPU load) is peaking, that we usually run on 5 TaskManagers (with some peaks up to even 8). On low throughput, we are running the minimal number of just one TaskManager. This chart is showing nicely that Reactive Mode is working as expected in this experiment: The number of TaskManagers is adjusting to the load on the system.
+
+
+### Lessons Learned
```

Comment:
I think this section is way longer than it needs to be. I'd try to reduce it to around two paragraphs and be more concise: what was the problem you faced? What was the solution?

Reply:
It is indeed a pretty long section, but I want to explain to users want's going on, instead of just telling them that there was some problem. I tried to mitigate the problem by making the headline more specific -- this way people understand that they can skip if, if it is too specific for them.